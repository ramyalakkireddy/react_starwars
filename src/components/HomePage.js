import React, { useState, useEffect } from 'react';
import { makeStyles } from '@material-ui/core/styles';
import { connect } from 'react-redux';
import moment from 'moment';
import CssBaseline from '@material-ui/core/CssBaseline';
import { MenuItem, FormControl, Select} from '@material-ui/core';
import { List, ListItem, ListItemText} from '@material-ui/core';
import { AppBar, Toolbar, Typography, Grid, Divider, CircularProgress, Paper } from '@material-ui/core';
import { getCharacters, getMovies } from '../actions';
import logo from '../assests/star_wars_logo.png';
import poster from '../assests/starwars.jpg';

const useStyles = makeStyles((theme) => ({
  appBar: {
    alignItems: 'center',
    backgroundColor: '#000000',
  },
  spinner: {
    color: theme.palette.background.paper,
    height: '50px !important',
    width: '50px !important',
  },
  image: {
    height: '65px',
    width: '100px',
  },
  formControl: {
    margin: '8px 0',
    minWidth: 250,
    color: theme.palette.background.paper,
    border: '1px solid #ffffff',
  },
  selectText: {
    color: '#ffffff !important',
  },
  root: {
    height: '100vh',
    flexWrap: 'nowrap',
    backgroundImage: `url(${poster})`,
  },
  container: {
    display: 'flex',
    flexDirection: 'column',
    width: 'fit-content',
    margin: '32px 32px 8px 32px;',
    color: theme.palette.background.paper,
  },
  demo: {
   minWidth: 250,
 },
 label: {
   fontWeight: 'bold',
 }
}));

const HomePage = (props) => {
  const classes = useStyles();
  const [character, setCharacter] = useState('');
  const [lastMovie, setLastMovie] = useState('');

  const { characterList, moviesList, loading } = props;

  useEffect(() => {
    props.getCharacters();
  }, []);

  useEffect(() => {
    props.getMovies(character);
  }, [character]);

  useEffect(() => {
    getLastMovie();
  }, [props.moviesList]);

  const handleChange = (event) => {
    setCharacter(event.target.value);
  };

  const getLastMovie = () => {
    let sortedList = [];
    if(moviesList.length > 0) {
      sortedList =  moviesList.reduce((a, b) => {
          return new Date(a.release_date) > new Date(b.release_date) ? a : b;
        });
      setLastMovie(sortedList.title + ' ('+ moment(sortedList.release_date).format('YYYY')+')');
      }
  }

  return (
    <React.Fragment>
      <CssBaseline />
      <AppBar position="static" className={classes.appBar}>
        <Toolbar>
          <img src={logo} className={classes.image} alt="starwars-logo" />
        </Toolbar>
      </AppBar>
      <Divider />
      <Paper elevation={3}>
        {loading ?
          <Grid container
            justify="center"
            className={classes.root}
          >
          <CircularProgress className={classes.spinner} /></Grid> :
          <div className={classes.root}>
            <Grid container
              justify="center"
            >
              <Grid container item xs={9} md={9} sm={9} lg={3} className={classes.container}>
                <Typography variant="h5" className={classes.label} display="inline">
                  Select Character:
                </Typography>
                <FormControl variant="outlined" className={classes.formControl}>
                  <Select
                    labelId="demo-simple-select-outlined-label"
                    id="demo-simple-select-outlined"
                    value={character}
                    onChange={handleChange}
                    inputProps={{
                      classes: {
                        root: classes.selectText,
                        icon: classes.selectText,
                      },
                    }}
                  >
                  {characterList.length > 0 && characterList.map((character) => {
                    return (
                      <MenuItem value={character.name} key={character.name}>{character.name}</MenuItem>
                      );
                    })
                  }
                  </Select>
                </FormControl>
              </Grid>
            {moviesList.length > 0 &&
              <Grid container item xs={9} md={9} sm={9} lg={3} className={classes.container}>
                <Typography variant="h5" className={classes.label}>
                  List of Movies:
                </Typography>
                <div className={classes.demo}>
                  <List>
                  {moviesList.map((movie) => {
                    return (
                      <ListItem key={movie.title} button>
                        <ListItemText primary={movie.title} />
                      </ListItem>
                    );
                  })
                  }
                  </List>
                </div>
              </Grid>}
              {moviesList.length > 0 && <Grid container item xs={9} md={9} sm={9} lg={3} className={classes.container}>
                <Typography variant="h5" className={classes.label}>
                  Name/Year last Movie:
                </Typography>
                <p>{lastMovie}</p>
              </Grid>}
            </Grid>
          </div>}
        </Paper>
      </React.Fragment>
  );
}

const mapStateToProps = state => ({
  characterList: state.characterList,
  moviesList: state.moviesList,
  loading: state.loading
});

const mapDispatchToProps = {
  getCharacters,
  getMovies
};

export default connect(mapStateToProps, mapDispatchToProps)(HomePage);
