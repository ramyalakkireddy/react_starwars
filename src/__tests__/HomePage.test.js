import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import configureStore from "redux-mock-store";
import HomePage from '../components/HomePage';
import "../setupTests";

it('should render correctly <HomePage>', () => {

  const mockStore = configureStore();
  const store = mockStore({
           characterList: [{
                 "name": "Luke Skywalker",
                 "height": "172",
                 "mass": "77",
                 "hair_color": "blond",
                 "skin_color": "fair",
                 "eye_color": "blue",
                 "birth_year": "19BBY",
                 "gender": "male",
                 "homeworld": "http://swapi.dev/api/planets/1/",
                 "films": [
                     "http://swapi.dev/api/films/1/",
                     "http://swapi.dev/api/films/2/",
                     "http://swapi.dev/api/films/3/",
                     "http://swapi.dev/api/films/6/"
                 ],
                 "species": [],
                 "vehicles": [
                     "http://swapi.dev/api/vehicles/14/",
                     "http://swapi.dev/api/vehicles/30/"
                 ],
                 "starships": [
                     "http://swapi.dev/api/starships/12/",
                     "http://swapi.dev/api/starships/22/"
                 ],
                 "created": "2014-12-09T13:50:51.644000Z",
                 "edited": "2014-12-20T21:17:56.891000Z",
                 "url": "http://swapi.dev/api/people/1/"
             }],
          moviesList: [
            {
              "title": "A New Hope",
              "episode_id": 4,
              "opening_crawl": "It is a period of civil war.\r\nRebel spaceships, striking\r\nfrom a hidden base, have won\r\ntheir first victory against\r\nthe evil Galactic Empire.\r\n\r\nDuring the battle, Rebel\r\nspies managed to steal secret\r\nplans to the Empire's\r\nultimate weapon, the DEATH\r\nSTAR, an armored space\r\nstation with enough power\r\nto destroy an entire planet.\r\n\r\nPursued by the Empire's\r\nsinister agents, Princess\r\nLeia races home aboard her\r\nstarship, custodian of the\r\nstolen plans that can save her\r\npeople and restore\r\nfreedom to the galaxy....",
              "director": "George Lucas",
              "producer": "Gary Kurtz, Rick McCallum",
              "release_date": "1977-05-25",
              "characters": [
                  "http://swapi.dev/api/people/1/",
                  "http://swapi.dev/api/people/2/",
                  "http://swapi.dev/api/people/3/",
                  "http://swapi.dev/api/people/4/",
                  "http://swapi.dev/api/people/5/",
                  "http://swapi.dev/api/people/6/",
                  "http://swapi.dev/api/people/7/",
                  "http://swapi.dev/api/people/8/",
                  "http://swapi.dev/api/people/9/",
                  "http://swapi.dev/api/people/10/",
                  "http://swapi.dev/api/people/12/",
                  "http://swapi.dev/api/people/13/",
                  "http://swapi.dev/api/people/14/",
                  "http://swapi.dev/api/people/15/",
                  "http://swapi.dev/api/people/16/",
                  "http://swapi.dev/api/people/18/",
                  "http://swapi.dev/api/people/19/",
                  "http://swapi.dev/api/people/81/"
              ],
              "planets": [
                  "http://swapi.dev/api/planets/1/",
                  "http://swapi.dev/api/planets/2/",
                  "http://swapi.dev/api/planets/3/"
              ],
              "starships": [
                  "http://swapi.dev/api/starships/2/",
                  "http://swapi.dev/api/starships/3/",
                  "http://swapi.dev/api/starships/5/",
                  "http://swapi.dev/api/starships/9/",
                  "http://swapi.dev/api/starships/10/",
                  "http://swapi.dev/api/starships/11/",
                  "http://swapi.dev/api/starships/12/",
                  "http://swapi.dev/api/starships/13/"
              ],
              "vehicles": [
                  "http://swapi.dev/api/vehicles/4/",
                  "http://swapi.dev/api/vehicles/6/",
                  "http://swapi.dev/api/vehicles/7/",
                  "http://swapi.dev/api/vehicles/8/"
              ],
              "species": [
                  "http://swapi.dev/api/species/1/",
                  "http://swapi.dev/api/species/2/",
                  "http://swapi.dev/api/species/3/",
                  "http://swapi.dev/api/species/4/",
                  "http://swapi.dev/api/species/5/"
              ],
              "created": "2014-12-10T14:23:31.880000Z",
              "edited": "2014-12-20T19:49:45.256000Z",
              "url": "http://swapi.dev/api/films/1/"
          }],
          loading: false
        });
  const wrapper = mount(
          <Provider store={store}>
              <HomePage />
          </Provider>);
  expect(wrapper).toMatchSnapshot();
});
