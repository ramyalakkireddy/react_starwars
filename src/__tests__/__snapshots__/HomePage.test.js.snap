// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly <HomePage> 1`] = `ReactWrapper {}`;
